core/market/tests/test_proposal.rs

```diff
@@ -78,3 +82,68 @@ async fn test_get_proposal_not_found() {
         result
     );
 }
+
+/// We don't want to give advantage for the oldest Offers, so we should shuffle
+/// results of `collect_offers` endpoint.
+#[cfg_attr(not(feature = "test-suite"), ignore)]
+#[serial_test::serial]
+async fn test_proposal_random_shuffle() {
+    let mut network = MarketsNetwork::new(None)
+        .await
+        .add_market_instance("Node-1")
+        .await;
+
+    let num = 10;
+    for i in 0..num {
+        network = network
+            .add_market_instance(&format!("Provider-{}", i))
+            .await;
+    }
+
+    let market1 = network.get_market("Node-1");
+    let identity1 = network.get_default_id("Node-1");
+
+    let demand_id = market1
+        .subscribe_demand(&sample_demand(), &identity1)
+        .await
+        .unwrap();
+
+    // Wait between subscribing Offers. Thanks to this, Offers
+    // will be propagated and added to queue in order.
+    let mut offers = vec![];
+    let mut ids = vec![];
+    for i in 0..num {
+        let node_name = format!("Provider-{}", i);
+        let market = network.get_market(&node_name);
+        let identity = network.get_default_id(&node_name);
+
+        offers.push(
+            market
+                .subscribe_offer(&sample_offer(), &identity)
+                .await
+                .unwrap(),
+        );
+        ids.push(identity.identity);
+
+        tokio::time::delay_for(Duration::from_millis(200)).await;
+    }
+
+    let events = market1
+        .query_events(&demand_id, 1.2, Some(num + 4))
+        .await
+        .unwrap();
+    assert_eq!(events.len(), num as usize);
+
+    let incoming_ids = events
+        .into_iter()
+        .filter_map(|event| match event {
+            RequestorEvent::ProposalEvent { proposal, .. } => Some(proposal.issuer_id),
+            _ => None,
+        })
+        .collect::<Vec<NodeId>>();
+
+    // If proposals were really shuffled, we expect incoming order to be different
+    // from initialization order.
+    assert_eq!(incoming_ids.len(), ids.len());
+    assert_ne!(incoming_ids, ids);
```

Comment:
buggy. it will sometimes match ;)

Reply:
I can increase number of Offers to minimize probability ;)

Reply:
;) that's even worse, leave for now until first fail